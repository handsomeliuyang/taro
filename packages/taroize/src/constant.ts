export const reserveKeyWords = new Set([
  'break',
  'case',
  'catch',
  'continue',
  'default',
  'delete',
  'do',
  'else',
  'finally',
  'for',
  'function',
  'if',
  'in',
  'instanceof',
  'new',
  'return',
  'switch',
  'this',
  'throw',
  'try',
  'typeof',
  'var',
  'void',
  'while',
  'with',
  'abstract',
  'boolean',
  'byte',
  'char',
  'class',
  'const',
  'debugger',
  'double',
  'enum',
  'export',
  'extends',
  'final',
  'float',
  'goto',
  'implements',
  'import',
  'int',
  'interface',
  'long',
  'native',
  'package',
  'private',
  'protected',
  'public',
  'short',
  'static',
  'super',
  'synchronized',
  'throws',
  'transient',
])

export const InternalKeyWorld = new Set([
  'Array',
  'Object',
  'String',
  'Number',
  'Boolean',
  'Function',
  'Date',
  'RegExp',
  'Math',
  'Error',
  'JSON',
  'Promise',
  'Map',
  'Symbol',
  'Reflect',
])

export const navigateFunc = new Set(['navigateTo', 'redirectTo', 'reLaunch'])
