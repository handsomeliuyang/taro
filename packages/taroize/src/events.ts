export const specialEvents = new Map<string, string>()
specialEvents.set('bindtimeupdate', 'onTimeUpdate')
specialEvents.set('bindgetphoneNumber', 'onGetPhoneNumber')
specialEvents.set('bindchooseavatar', 'onChooseAvatar')
specialEvents.set('bindgetrealnameauthinfo', 'onGetRealnameAuthInfo')
specialEvents.set('bindopensetting', 'onOpenSetting')
specialEvents.set('bindscancode', 'onScanCode')
specialEvents.set('bindstatechange', 'onStateChange')
specialEvents.set('bindhtouchmove', 'onHTouchMove')
specialEvents.set('bindvtouchmove', 'onVTouchMove')
specialEvents.set('bindcolumnchange', 'onColumnChange')
specialEvents.set('bindscrolltoupper', 'onScrollToUpper')
specialEvents.set('bindscrolltolower', 'onScrollToLower')
specialEvents.set('bindanimationfinish', 'onAnimationFinish')
specialEvents.set('bindfullscreenchange', 'onFullscreenChange')
specialEvents.set('bindtouchstart', 'onTouchStart')
specialEvents.set('bindtouchmove', 'onTouchMove')
specialEvents.set('bindtouchcancel', 'onTouchCancel')
specialEvents.set('bindtouchend', 'onTouchEnd')
specialEvents.set('bindlongpress', 'onLongPress')
specialEvents.set('bindlongclick', 'onLongClick')
specialEvents.set('bindtransitionend', 'onTransitionEnd')
specialEvents.set('bindanimationstart', 'onAnimationStart')
specialEvents.set('bindanimationtteration', 'onAnimationIteration')
specialEvents.set('bindanimationend', 'onAnimationEnd')
specialEvents.set('bindtouchforcechange', 'onTouchForceChange')
specialEvents.set('bindtap', 'onClick')
specialEvents.forEach((value, key) => {
  specialEvents.set(key.replace(/^bind/, 'catch'), value)
})

export const specialSameEvents = new Map<string, string>()
specialSameEvents.set('bind:timeupdate', 'onTimeUpdate')
specialSameEvents.set('bind:getphoneNumber', 'onGetPhoneNumber')
specialSameEvents.set('bind:chooseavatar', 'onChooseAvatar')
specialSameEvents.set('bind:getrealnameauthinfo', 'onGetRealnameAuthInfo')
specialSameEvents.set('bind:opensetting', 'onOpenSetting')
specialSameEvents.set('bind:scancode', 'onScanCode')
specialSameEvents.set('bind:statechange', 'onStateChange')
specialSameEvents.set('bind:htouchmove', 'onHTouchMove')
specialSameEvents.set('bind:vtouchmove', 'onVTouchMove')
specialSameEvents.set('bind:columnchange', 'onColumnChange')
specialSameEvents.set('bind:scrolltoupper', 'onScrollToUpper')
specialSameEvents.set('bind:scrolltolower', 'onScrollToLower')
specialSameEvents.set('bind:animationfinish', 'onAnimationFinish')
specialSameEvents.set('bind:fullscreenchange', 'onFullscreenChange')
specialSameEvents.set('bind:touchstart', 'onTouchStart')
specialSameEvents.set('bind:touchmove', 'onTouchMove')
specialSameEvents.set('bind:touchcancel', 'onTouchCancel')
specialSameEvents.set('bind:touchend', 'onTouchEnd')
specialSameEvents.set('bind:longpress', 'onLongPress')
specialSameEvents.set('bind:longclick', 'onLongClick')
specialSameEvents.set('bind:transitionend', 'onTransitionEnd')
specialSameEvents.set('bind:animationstart', 'onAnimationStart')
specialSameEvents.set('bind:animationtteration', 'onAnimationIteration')
specialSameEvents.set('bind:animationend', 'onAnimationEnd')
specialSameEvents.set('bind:touchforcechange', 'onTouchForceChange')
specialSameEvents.set('bind:tap', 'onClick')
specialSameEvents.forEach((value, key) => {
  specialSameEvents.set(key.replace(/^bind/, 'catch'), value)
})