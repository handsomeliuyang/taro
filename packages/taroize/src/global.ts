export const usedComponents = new Set<string>()

export const errors: string[] = []

export const globals = {
  hasCatchTrue: false,
  logFilePath: '',
}

export const THIRD_PARTY_COMPONENTS = new Set<string>()

export const resetGlobals = (logFilePath) => {
  globals.hasCatchTrue = false
  globals.logFilePath = logFilePath
  THIRD_PARTY_COMPONENTS.clear()
}
