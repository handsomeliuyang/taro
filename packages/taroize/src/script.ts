import traverse, { NodePath, Visitor } from '@babel/traverse'
import * as t from '@babel/types'

import { usedComponents } from './global'
import { buildBlockElement, buildImportStatement, buildRender, codeFrameError, parseCode } from './utils'
import { WXS } from './wxml'

const defaultClassName = '_C'

const buildDecorator = (id: t.Identifier | t.ObjectExpression, isApp = false) => {
  const args: any[] = [id]
  isApp && args.push(t.booleanLiteral(true))
  return t.decorator(t.callExpression(t.identifier('withWeapp'), args))
}

export function replaceIdentifier (callee: NodePath<t.Node>) {
  if (callee.isIdentifier()) {
    const name = callee.node.name
    if (name === 'getApp' || name === 'getCurrentPages') {
      callee.replaceWith(t.memberExpression(t.identifier('Taro'), callee.node))
    }
  }
}

// babel 6升级babel 7适配
export function replaceMemberExpression (callee: NodePath<t.Node>) {
  if (callee.isMemberExpression()) {
    const object = callee.get('object') as NodePath<t.Identifier>
    if (t.isIdentifier(object.node, { name: 'wx' })) {
      object.replaceWith(t.identifier('Taro'))
    }
  }
}

export function parseScript (
  script?: string,
  scriptPath?: string,
  returned?: t.Expression,
  wxses: WXS[] = [],
  refId?: Set<string>,
  isApp = false
) {
  script = script || 'Page({})'
  if (t.isJSXText(returned as any)) {
    const block = buildBlockElement()
    block.children = [returned as any]
    returned = block
  }
  let ast = parseCode(script, scriptPath as string)
  let classDecl!: t.ClassDeclaration
  let foundWXInstance = false
  const vistor: Visitor = {
    BlockStatement (path) {
      path.scope.rename('wx', 'Taro')
    },
    Identifier (path) {
      if (path.isReferenced() && path.node.name === 'wx') {
        path.replaceWith(t.identifier('Taro'))
      }
    },
    CallExpression (path) {
      const callee = path.get('callee')
      replaceIdentifier(callee as NodePath<t.Node>)
      replaceMemberExpression(callee as NodePath<t.Node>)
      if (
        callee.isIdentifier({ name: 'Page' }) ||
        callee.isIdentifier({ name: 'Component' }) ||
        callee.isIdentifier({ name: 'App' })
      ) {
        foundWXInstance = true
        const componentType = callee.node.name
        classDecl = parsePage(path, returned || t.nullLiteral(), componentType, refId, wxses, isApp)
        ast.program.body.push(
          classDecl,
          t.exportDefaultDeclaration(t.identifier(componentType !== 'App' ? defaultClassName : 'App'))
        )
        // path.insertAfter(t.exportDefaultDeclaration(t.identifier(defaultClassName)))
        path.remove()
      }
    },
    VariableDeclaration (path) {
      const variableDeclarator = path.node.declarations[0]
      variableDeclarator.init
      if (t.isVariableDeclarator(variableDeclarator)) {
        // 去除commonjs转es6生成的var module = {exports: {},}
        if (
          t.isIdentifier(variableDeclarator.id) &&
          variableDeclarator.id.name === 'module' &&
          t.isObjectExpression(variableDeclarator.init)
        ) {
          const init: t.ObjectExpression = variableDeclarator.init
          if (
            init.properties &&
            init.properties.length > 0 &&
            t.isObjectProperty(init.properties[0]) &&
            t.isIdentifier(init.properties[0].key) &&
            init.properties[0].key.name === 'exports'
          ) {
            path.remove()
          }
        }
        // 去除commonjs转es6生成的var exports = module.exports
        if (
          t.isIdentifier(variableDeclarator.id) &&
          variableDeclarator.id.name === 'exports' &&
          t.isMemberExpression(variableDeclarator.init)
        ) {
          const init: t.MemberExpression = variableDeclarator.init
          if (
            t.isIdentifier(init.object) &&
            init.object.name === 'module' &&
            t.isIdentifier(init.property) &&
            init.property.name === 'exports'
          ) {
            path.remove()
          }
        }
      }
    },
    ExportDefaultDeclaration (path) {
      const declaration = path.node.declaration
      // 去除commonjs转es6生成的export default module.exports
      if (t.isMemberExpression(declaration)) {
        if (
          t.isIdentifier(declaration.object) &&
          declaration.object.name === 'module' &&
          t.isIdentifier(declaration.property) &&
          declaration.property.name === 'exports'
        ) {
          path.remove()
        }
      }
    },
  }

  traverse(ast, vistor)

  if (!foundWXInstance) {
    ast = parseCode(script + ';Component({})')
    traverse(ast, vistor)
  }

  const taroComponentsImport = buildImportStatement('@tarojs/components', [...usedComponents])
  const taroImport = buildImportStatement('@tarojs/taro', [], 'Taro')
  const reactImport = buildImportStatement('react', [], 'React')
  const withWeappImport = buildImportStatement('@tarojs/with-weapp', [], 'withWeapp')
  ast.program.body.unshift(
    taroComponentsImport,
    reactImport,
    taroImport,
    withWeappImport,
    ...wxses.filter((wxs) => !wxs.src.startsWith('./wxs__')).map((wxs) => buildImportStatement(wxs.src, [], wxs.module))
  )

  return ast
}

function parsePage (
  pagePath: NodePath<t.CallExpression>,
  returned: t.Expression,
  componentType?: string,
  refId?: Set<string>,
  wxses?: WXS[],
  isApp = false
) {
  const stateKeys: string[] = []
  pagePath.traverse({
    CallExpression (path) {
      const callee = path.get('callee')
      replaceIdentifier(callee as NodePath<t.Node>)
      replaceMemberExpression(callee as NodePath<t.Node>)
    },
  })
  if (refId) {
    refId.forEach((id) => {
      if (!stateKeys.includes(id)) {
        stateKeys.push(id)
      }
    })
  }
  const propsKeys: string[] = []
  const arg = pagePath.get('arguments')[0]

  const classBody: any[] = []
  if (arg.isObjectExpression() || arg.isIdentifier()) {
    //
  } else {
    throw codeFrameError(arg.node, `${componentType || '组件'} 的第一个参数必须是一个对象或变量才能转换。`)
  }

  const wxsNames = new Set(wxses ? wxses.map((w) => w.module) : [])

  const renderFunc = buildRender(
    componentType === 'App'
      ? t.memberExpression(t.memberExpression(t.thisExpression(), t.identifier('props')), t.identifier('children'))
      : returned,
    stateKeys.filter((s) => !wxsNames.has(s)),
    propsKeys
  )

  const classDecl = t.classDeclaration(
    t.identifier(componentType === 'App' ? 'App' : defaultClassName),
    t.memberExpression(t.identifier('React'), t.identifier('Component')),
    t.classBody(classBody.concat(renderFunc)),
    []
  )

  classDecl.decorators = [buildDecorator(arg.node, isApp)]

  return classDecl
}
