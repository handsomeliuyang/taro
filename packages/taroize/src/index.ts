import * as t from '@babel/types'
import { promoteRelativePath } from '@tarojs/helper'
import * as path from 'path'

import { errors, resetGlobals, THIRD_PARTY_COMPONENTS } from './global'
import { parseScript } from './script'
import { setting } from './utils'
import { parseVue } from './vue'
import { parseWXML } from './wxml'

interface Option {
  json?: string
  script?: string
  scriptPath?: string
  wxml?: string
  path: string
  rootPath: string
  framework: 'react' | 'vue'
  isApp?: boolean
}

export function parse (option: Option) {
  resetGlobals()
  setting.rootPath = option.rootPath
  if (option.json) {
    const config = JSON.parse(option.json)
    const usingComponents = config.usingComponents
    if (usingComponents) {
      for (const key in usingComponents) {
        if (usingComponents.hasOwnProperty(key)) {
          THIRD_PARTY_COMPONENTS.add(key)
        }
      }
    }
  }

  if (option.framework === 'vue') {
    const result = parseVue(option.path, option.wxml || '', option.script)
    return {
      ...result,
      errors,
    }
  }

  const { wxml, wxses, imports, refIds } = parseWXML(option.path, option.wxml)
  setting.sourceCode = option.script!
  const ast = parseScript(option.script, option.scriptPath, wxml as t.Expression, wxses, refIds, option.isApp)

  // 添加 cacheOptions 引入语句
  if ( option.scriptPath ) {
    const currentPath = option.scriptPath || ''
    const cacheOptionsPath = path.resolve(option.rootPath, 'utils', '_cacheOptions.js')
    const importCacheOptionsUrl = promoteRelativePath(path.relative(currentPath, cacheOptionsPath))
    const importCacheOptionsAst = t.importDeclaration([
      t.importSpecifier(t.identifier('cacheOptions'), t.identifier('cacheOptions'))
    ], t.stringLiteral(importCacheOptionsUrl))
    ast.program.body.unshift(importCacheOptionsAst)
  }

  return {
    ast,
    imports,
    errors,
  }
}
