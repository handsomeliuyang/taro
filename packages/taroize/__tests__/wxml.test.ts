import * as t from '@babel/types'

import { convertStyleUnit,parseContent, parseStyle, parseWXML } from '../src/wxml'

interface Option {
  path: string
  rootPath: string
  framework: 'react' | 'vue'
  json?: string
  script?: string
  scriptPath?: string
  wxml?: string
  isApp?: boolean
}

const option: Option = {
  framework: 'react',
  json: '{}',
  path: '',
  rootPath: '',
  script: '',
  scriptPath: '',
  wxml: ''
}

describe('wxml.ts测试', () => {
  test('当wxml 是 undefined', () => {
    option.wxml = undefined
    //  parseWXML会先获取缓存，所有每个用例的path需要保持其唯一性
    option.path = 'wxml_undefined'
    const { wxml, wxses, imports, refIds } = parseWXML(option.path, option.wxml)
    expect(wxml?.type).toBe('NullLiteral')
    expect(wxses).toEqual([])
    expect(imports).toEqual([])
    expect(Array.from(refIds)).toEqual(Array.from(new Set()))
  })

  test('wxml 简单应用', () => {
    option.wxml = `<view>Hello Word!</view>`
    //  parseWXML会先获取缓存，所有每个用例的path需要保持其唯一性
    option.path = 'wxml__'
    const { wxml } = parseWXML(option.path, option.wxml)
    // wxml转为ast
    expect(wxml).toBeTruthy()
    expect(wxml?.type).toBe('JSXElement')
  })

  test('wxml 中存在 wxs', () => {
    option.wxml = `<wxs module="wxs_demo">
                        module.exports = {
                            data: 'wxs demo'
                        }
                    </wxs>
                    <view>Hello Word!</view>
                    <view>{{wxs_demo.data}}</view>`
    //  parseWXML会先获取缓存，所有每个用例的path需要保持其唯一性
    option.path = 'wxml_wxs'
    const { wxses, imports } = parseWXML(option.path, option.wxml)
    expect(wxses).toMatchSnapshot()
    expect(imports).toMatchSnapshot()
  })
})

describe('parseContent', () => {
  test('节点key的值解析', () => {
    const contentInput = `{{list}}`
    const { type, content } = parseContent(contentInput)
    expect(type).toBe('expression')
    expect(content).toBe(`(list)`)
  })
})

describe('style属性的解析', () => {
  // 第一种以style="xxx:xxx"的用法
  test('style = xxx:xxx', () => {
    let contentInput = 'color: red;background-color: aqua;'
    contentInput = convertStyleUnit(contentInput)
    const styleParseReslut = parseStyle('style', contentInput)
    if (t.isJSXAttribute(styleParseReslut)) {
      expect(styleParseReslut.type).toEqual('JSXAttribute')
    } else {
      const { type, content } = styleParseReslut
      expect(type).toBe('expression')
      expect(content).toBe('(color: red;background-color: aqua;)')
    }
  })

  test('style = xxx:{{ xxx }}', () => {
    let contentInput = 'color: red;background-color: aqua;font-size: {{ fontSize }};'
    contentInput = convertStyleUnit(contentInput)
    const styleParseReslut = parseStyle('style', contentInput)
    if (t.isJSXAttribute(styleParseReslut)) {
      expect(styleParseReslut.type).toEqual('JSXAttribute')
    } else {
      const { type, content } = styleParseReslut
      expect(type).toBe('expression')
      expect(content).toBe('(color: red;background-color: aqua;font-size: {{ fontSize }};)')
    }
  })

  test('style = xxx:{{ xxx }}px', () => {
    let contentInput = 'height: {{ height }}px;width: 100px;background-color: red;'
    contentInput = convertStyleUnit(contentInput)
    const styleParseReslut = parseStyle('style', contentInput)
    if (t.isJSXAttribute(styleParseReslut)) {
      expect(styleParseReslut.type).toEqual('JSXAttribute')
    } else {
      const { type, content } = styleParseReslut
      expect(type).toBe('expression')
      expect(content).toBe('(height: {{ height }}px;width: 5rem;background-color: red;)')
    }
  })

  test('style = xxx:{{ xxx }}, 含有三元运算符', () => {
    let contentInput = 'height: {{ list.length > 10 ? 200 : 100 }}px;width: 100px;background-color: green;'
    contentInput = convertStyleUnit(contentInput)
    const styleParseReslut = parseStyle('style', contentInput)
    if (t.isJSXAttribute(styleParseReslut)) {
      expect(styleParseReslut.type).toEqual('JSXAttribute')
    } else {
      const { type, content } = styleParseReslut
      expect(type).toBe('expression')
      expect(content).toBe('(height: {{ list.length > 10 ? 200 : 100 }}px;width: 100px;background-color: green);')
    }
  })

  test('style = xxx:{{ xxx + 10 }}px', () => {
    let contentInput = 'height: {{ height + 10 }}px;width: 100px;background-color: red;'
    contentInput = convertStyleUnit(contentInput)
    const styleParseReslut = parseStyle('style', contentInput)
    if (t.isJSXAttribute(styleParseReslut)) {
      expect(styleParseReslut.type).toEqual('JSXAttribute')
    } else {
      const { type, content } = styleParseReslut
      expect(type).toBe('expression')
      expect(content).toBe('({{ height + 10 }}px;width: 5rem;background-color: red;)')
    }
  })

  test('style = xxx:xxx混用情况1', () => {
    let contentInput = 'height: {{ height }}px;width: {{ width }}px;background-color: {{ color }};'
    contentInput = convertStyleUnit(contentInput)
    const styleParseReslut = parseStyle('style', contentInput)
    if (t.isJSXAttribute(styleParseReslut)) {
      expect(styleParseReslut.type).toEqual('JSXAttribute')
    } else {
      const { type, content } = styleParseReslut
      expect(type).toBe('expression')
      expect(content).toBe('(height: {{ height }}px;width: {{ width }}px;background-color: {{ color }})')
    }
  })

  test('style = xxx:xxx混用情况2', () => {
    let contentInput = 'height: {{ height }}px;width: {{ width }}px;background-color: {{ false ? color : "blue" }};'
    contentInput = convertStyleUnit(contentInput)
    const styleParseReslut = parseStyle('style', contentInput)
    if (t.isJSXAttribute(styleParseReslut)) {
      expect(styleParseReslut.type).toEqual('JSXAttribute')
    } else {
      const { type, content } = styleParseReslut
      expect(type).toBe('expression')
      expect(content).toBe('(height: {{ height }}px;width: {{ width }}px;background-color: {{ false ? color : "blue" }};)')
    }
  })

  // 第二种以双括号用法
  test('style = {{ xxx }}变量形式', () => {
    let contentInput = '{{ myStyle }}'
    contentInput = convertStyleUnit(contentInput)
    const styleParseReslut = parseStyle('style', contentInput)
    if (t.isJSXAttribute(styleParseReslut)) {
      expect(styleParseReslut).toEqual(t.jSXAttribute)
    } else {
      const { type, content } = styleParseReslut
      expect(type).toBe('expression')
      expect(content).toBe('(myStyle)')
    }
  })

  test('style = {{ xxx }}字符串形式', () => {
    let contentInput = "{{ 'height: 100px;width: 100px;background-color: red;' }}"
    contentInput = convertStyleUnit(contentInput)
    const styleParseReslut = parseStyle('style', contentInput)
    if (t.isJSXAttribute(styleParseReslut)) {
      expect(styleParseReslut).toEqual(t.jSXAttribute)
    } else {
      const { type, content } = styleParseReslut
      expect(type).toBe('expression')
      expect(content).toBe("('height: 5rem;width: 5rem;background-color: red;')")
    }
  })

  test('style = {{ xxx }}含三元运算符', () => {
    let contentInput = "{{ index > 1 ? 'height: 100px;width: 100px;background-color: green;' : '' }}"
    contentInput = convertStyleUnit(contentInput)
    const styleParseReslut = parseStyle('style', contentInput)
    if (t.isJSXAttribute(styleParseReslut)) {
      expect(styleParseReslut).toEqual(t.jSXAttribute)
    } else {
      const { type, content } = styleParseReslut
      expect(type).toBe('expression')
      expect(content).toBe("(index > 1 ? 'height: 5rem;width: 5rem;background-color: green;' : '')")
    }
  })

  test('style = {{ xxx }}含三元运算符和变量', () => {
    let contentInput = '{{ index ? "background-color: " + backgroundColor + ";" : "" }}'
    contentInput = convertStyleUnit(contentInput)
    const styleParseReslut = parseStyle('style', contentInput)
    if (t.isJSXAttribute(styleParseReslut)) {
      expect(styleParseReslut).toEqual(t.jSXAttribute)
    } else {
      const { type, content } = styleParseReslut
      expect(type).toBe('expression')
      expect(content).toBe('(index ? "background-color: " + backgroundColor + ";" : "")')
    }
  })

  // 第三种xxx:xxx和{{ xxx }}混用情况
  test('xxx:xxx和{{ xxx }}混用情况1', () => {
    let contentInput = 'background-color: aqua;{{ global_css }}'
    contentInput = convertStyleUnit(contentInput)
    const styleParseReslut = parseStyle('style', contentInput)
    if (t.isJSXAttribute(styleParseReslut)) {
      expect(styleParseReslut).toEqual(t.jSXAttribute)
    } else {
      const { type, content } = styleParseReslut
      expect(type).toBe('expression')
      expect(content).toBe('"background-color: aqua;"+(global_css)')
    }
  })

  test('xxx:xxx和{{ xxx }}混用情况2', () => {
    let contentInput = 'background-color: aqua;{{ true ? global_css : "" }}'
    contentInput = convertStyleUnit(contentInput)
    const styleParseReslut = parseStyle('style', contentInput)
    if (t.isJSXAttribute(styleParseReslut)) {
      expect(styleParseReslut).toEqual(t.jSXAttribute)
    } else {
      const { type, content } = styleParseReslut
      expect(type).toBe('expression')
      expect(content).toBe('"background-color: aqua;"+(true ? global_css : "")')
    }
  })

  test('xxx:xxx和{{ xxx }}混用情况3', () => {
    let contentInput = 'background-color: {{ color }};{{ global_css }}'
    contentInput = convertStyleUnit(contentInput)
    const styleParseReslut = parseStyle('style', contentInput)
    if (t.isJSXAttribute(styleParseReslut)) {
      expect(styleParseReslut).toEqual(t.jSXAttribute)
    } else {
      const { type, content } = styleParseReslut
      expect(type).toBe('expression')
      expect(content).toBe('"background-color: "+(color)+";"+(global_css)')
    }
  })

  test('xxx:xxx和{{ xxx }}混用情况4', () => {
    let contentInput = 'background-color: {{ true ? color : "" }};{{ global_css }}'
    contentInput = convertStyleUnit(contentInput)
    const styleParseReslut = parseStyle('style', contentInput)
    if (t.isJSXAttribute(styleParseReslut)) {
      expect(styleParseReslut).toEqual(t.jSXAttribute)
    } else {
      const { type, content } = styleParseReslut
      expect(type).toBe('expression')
      expect(content).toBe('"background-color: "+(true ? color : "")+";"+(global_css)')
    }
  })

  test('xxx:xxx和{{ xxx }}混用情况5', () => {
    let contentInput = '{{ global_css }}{{ myStyle }}'
    contentInput = convertStyleUnit(contentInput)
    const styleParseReslut = parseStyle('style', contentInput)
    if (t.isJSXAttribute(styleParseReslut)) {
      expect(styleParseReslut).toEqual(t.jSXAttribute)
    } else {
      const { type, content } = styleParseReslut
      expect(type).toBe('expression')
      expect(content).toBe('(global_css)+(myStyle)')
    }
  })

  // 第四种情况,与第一种情况有差别
  test('style = xxx:xxx, 但是key是变量', () => {
    let contentInput = '{{ styleName }}: red;height: 100px;width: 100px;'
    contentInput = convertStyleUnit(contentInput)
    const styleParseReslut = parseStyle('style', contentInput)
    if (t.isJSXAttribute(styleParseReslut)) {
      expect(styleParseReslut).toEqual(t.jSXAttribute)
    } else {
      const { type, content } = styleParseReslut
      expect(type).toBe('expression')
      expect(content).toBe('(styleName)+": red;height: 5rem;width: 5rem;"')
    }
  })

  // style为空的特殊写法
  test('style为空的特殊写法1', () => {
    let contentInput = '{styleX}'
    contentInput = convertStyleUnit(contentInput)
    const styleParseReslut = parseStyle('style', contentInput)
    if (t.isJSXAttribute(styleParseReslut)) {
      expect(styleParseReslut).toEqual(t.jSXAttribute)
    } else {
      const { type, content } = styleParseReslut
      expect(type).toBe('raw')
      expect(content).toBe('{styleX}')
    }
  })

  test('style为空的特殊写法2', () => {
    let contentInput = '""'
    contentInput = convertStyleUnit(contentInput)
    const styleParseReslut = parseStyle('style', contentInput)
    if (t.isJSXAttribute(styleParseReslut)) {
      expect(styleParseReslut).toEqual(t.jSXAttribute)
    } else {
      const { type, content } = styleParseReslut
      expect(type).toBe('raw')
      expect(content).toBe('""')
    }
  })

  test('style为空的特殊写法3', () => {
    let contentInput = '{{}}'
    contentInput = convertStyleUnit(contentInput)
    const styleParseReslut = parseStyle('style', contentInput)
    if (t.isJSXAttribute(styleParseReslut)) {
      expect(styleParseReslut).toEqual(t.jSXAttribute)
    } else {
      const { type, content } = styleParseReslut
      expect(type).toBe('raw')
      expect(content).toBe('{{}}')
    }
  })

  test('style为空的特殊写法4', () => {
    let contentInput = '{{ "" }}'
    contentInput = convertStyleUnit(contentInput)
    const styleParseReslut = parseStyle('style', contentInput)
    if (t.isJSXAttribute(styleParseReslut)) {
      expect(styleParseReslut).toEqual(t.jSXAttribute)
    } else {
      const { type, content } = styleParseReslut
      expect(type).toBe('expression')
      expect(content).toBe('("")')
    }
  })
})
