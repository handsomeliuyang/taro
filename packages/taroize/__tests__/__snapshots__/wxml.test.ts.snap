// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wxml语法 wx:key="index"会转换成key="index" 1`] = `
data.map((item, index) => {
  return <View key={index}><Text>{item.name}</Text></View>;
})
`;

exports[`wxml语法 列表渲染 1`] = `
array.map((item, index) => {
  return <View>{index + ": " + item.message}</View>;
})
`;

exports[`wxs wxs 多行注释 1`] = `
module.exports = {
  foo: "'hello world' from tools.wxs"
  /*
    msg:'123',
    name:'xixi'
  */
};
`;

exports[`wxs wxs中 单行注释 1`] = `
module.exports = {
  foo: "'hello world' from tools.wxs"
  // msg:'123'
};
`;

exports[`wxs wxs标签中的getDate()转换为new Date() 1`] = `
Array [
  Object {
    module: wxs_getDate,
    src: ./wxs__wxs_getDate,
  },
]
`;

exports[`wxs wxs标签中的getDate()转换为new Date() 2`] = `
module.exports = {
  date1: new Date(),
  //参数为数字
  date2: new Date(1500000000000),
  //参数为字符串
  date3: new Date("2017-7-14")
};
`;

exports[`wxs wxs模块中的var regexp = getRegExp()转换为var regexp = new RegExp() 1`] = `
Array [
  Object {
    module: wxs_regexp,
    src: ./wxs__wxs_regexp,
  },
]
`;

exports[`wxs 页面中使用wxs 1`] = `
module.exports = {
  data: 'wxs demo'
};
`;

exports[`wx作用域属性解析 wx:for & wx:for-xxx 1`] = `
array.map((itemName, idx) => {
  return <View>{idx + ": " + itemName.message}</View>;
})
`;

exports[`wx作用域属性解析 wx:for & wx:if 1`] = `
modalKey.map((item, index) => {
  return <Block>{item.key && <View>{"the value is : " + item.value}</View>}</Block>;
})
`;

exports[`wx作用域属性解析 wx:if & wx:for 1`] = `
modalKey.map((item, index) => {
  return <Block>{item.key && <View>{"the value is : " + item.value}</View>}</Block>;
})
`;
